src/sparse/KokkosSparse_BlockCrsMatrix.hpp

```diff
@@ -0,0 +1,940 @@
+                 ordinal_type* const colidx__,
+                 const ordinal_type& blockDim,
+                 const ordinal_type& count) :
+    values_ (values), colidx_ (colidx__), blockDim_(blockDim), length (count)
+  {}
+
+  /// \brief Constructor with offset into \c colidx array
+  ///
+  /// \param values [in] Array of the row's values.
+  /// \param colidx [in] Array of the row's column indices.
+  /// \param blockDim [in] (Constant) stride between rows in
+  ///   within a block in the above arrays.
+  /// \param count [in] Number of blocks in the desired block-row
+  /// \param start [in] Offset into values and colidx of the desired block-row start.
+  ///   Note: The offset into the values array for a block-row equals
+  ///           num_blocks_prior_to_block-row*blockDim*blockDim
+  ///
+  /// \tparam OffsetType The type of \c start (see above).  Must be a
+  ///   built-in integer type.  This may differ from ordinal_type.
+  ///   For example, the matrix may have dimensions that fit in int,
+  ///   but a number of entries that does not fit in int.
+  template<class OffsetType>
+  KOKKOS_INLINE_FUNCTION
+  SparseBlockRowView (const typename MatrixType::values_type& values,
+                 const typename MatrixType::index_type& colidx__,
+                 const ordinal_type& blockDim,
+                 const ordinal_type& count,
+                 const OffsetType& start,
+                 const typename std::enable_if<std::is_integral<OffsetType>::value, int>::type& = 0) :
+    values_ (&values(start*blockDim*blockDim)), colidx_ (&colidx__(start)), blockDim_(blockDim), length (count)
+  {}
+
+  /// \brief Number of entries (i.e. blocks) in the row.
+  ///
+  /// This is a public const field rather than a public const method,
+  /// in order to avoid possible overhead of a method call if the
+  /// compiler is unable to inline that method call.
+  ///
+  /// We assume that rows contain no duplicate entries (i.e., entries
+  /// with the same column index).  Thus, a row may have up to
+  /// A.numCols() entries.  This means that the correct type of
+  /// 'length' is ordinal_type.
+  /// Here, length refers to the number of blocks in a block-row
+  const ordinal_type length;
+
+
+  // Return a pointer offset to full-row i of values_ array; 
+  //   user responsible for indexing into this pointer correctly
+  // Input: i must be the LOCAL row index within this block-row
+  // Output: pointer to values_ array at start of full row with local index i
+  KOKKOS_INLINE_FUNCTION
+  value_type* full_row_in_block_row (const ordinal_type& i) const {
+    return values_+(i*length*blockDim_) ;
+  }
+
+  // Return a pointer offset to local row i of block K of values_ array; 
+  //   user responsible for indexing into this pointer correctly
+  KOKKOS_INLINE_FUNCTION
+  value_type* local_row_in_block (const ordinal_type& K, const ordinal_type& i) const {
+    return (values_+(K*blockDim_ + i*length*blockDim_)) ;
+  }
+
```

Comment:
Minor thing. We probably need to comment that all pointer based interfaces are not guaranteed for backward compatibility as this type of interface is more used in kernels for performance not for ordinary users.

Reply:
Done. I made the commenting more consistent with other parts of the code also.